plugin/src/rules/max_statements.rs

```diff
@@ -0,0 +1,407 @@
+    }
+}
+
+#[derive(Default, Deserialize)]
+#[serde(default)]
+struct OptionsObject {
+    ignore_top_level_functions: bool,
+}
+
+struct Options {
+    max: usize,
+    ignore_top_level_functions: bool,
+}
+
+impl Options {
+    pub fn from_max_and_options_object(max: Max, options_object: OptionsObject) -> Self {
+        Self {
+            max: max.into(),
+            ignore_top_level_functions: options_object.ignore_top_level_functions,
+        }
+    }
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        OptionsVariants::default().into()
+    }
+}
+
+impl From<OptionsVariants> for Options {
+    fn from(value: OptionsVariants) -> Self {
+        match value {
+            OptionsVariants::EmptyList() => {
+                Self::from_max_and_options_object(Default::default(), Default::default())
+            }
+            OptionsVariants::JustMax(max) => {
+                Self::from_max_and_options_object(max[0], Default::default())
+            }
+            OptionsVariants::MaxAndOptionsObject(max, options_object) => {
+                Self::from_max_and_options_object(max, options_object)
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Options {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        Ok(OptionsVariants::deserialize(deserializer)?.into())
+    }
+}
+
+struct TopLevelFunction<'a> {
+    node: Node<'a>,
+    count: usize,
+}
+
+fn report_if_too_many_statements(
+    node: Node,
+    count: usize,
+    max: usize,
+    context: &QueryMatchContext,
+) {
+    if count <= max {
+        return;
+    }
+
+    let name = upper_case_first(&ast_utils::get_function_name_with_kind(node, context));
+
+    context.report(violation! {
+        node => node,
+        message_id => "exceed",
+        data => {
+            name => name,
+            count => count,
+            max => max,
+        }
+    });
+}
+
+pub fn max_statements_rule() -> Arc<dyn Rule> {
+    rule! {
+        name => "max-statements",
+        languages => [Javascript],
+        messages => [
+            exceed => "{{name}} has too many statements ({{count}}). Maximum allowed is {{max}}.",
+        ],
+        options_type => Options,
+        state => {
+            [per-config]
+            max_statements: usize = options.max,
+            ignore_top_level_functions: bool = options.ignore_top_level_functions,
+
+            [per-file-run]
+            function_stack: Vec<usize>,
+            top_level_functions: Vec<TopLevelFunction<'a>>,
+        },
+        listeners => [
+            r#"
+              (function_declaration) @c
+              (function) @c
+              (arrow_function) @c
+              (class_static_block) @c
+              (generator_function_declaration) @c
+              (generator_function) @c
+              (method_definition) @c
+            "# => |node, context| {
+                self.function_stack.push(0);
```

Comment:
I remember that initially I was avoiding supporting "node-exit" listeners so was achieving comparable behavior on a rule-by-rule basis with eg node stacks, but then I added node-exit listeners so it might make sense to go back through and use those instead?